### ISSUE TITLE
Proposal: Feature Reconstruction Loss for FutureState (Auxiliary Self-Supervised Task)

### ISSUE THREAD

**Issue by riita10069** [autowarefoundation/auto_e2e#13]
## Summary

The `FutureState` module currently outputs predicted future visual features (`[B, 1440, 7, 7] × 4`) during the forward pass, but these predictions are never used — there is no loss function that compares them against actual future observations. This issue proposes implementing the Feature Reconstruction Loss to activate FutureState as an auxiliary training signal and to enable introspection of model uncertainty.

## Current State

- `FutureState.forward()` predicts features at t+1.6s, t+3.2s, t+4.8s, t+6.4s
- `FrozenBackbone` exists in the codebase but is unused
- The predicted future features are returned from `AutoE2E.forward()` but discarded

### How it works

During training with sequential frame data, we compare what the model predicted the future would look like against what the future actually looked like:

```
Frame t (current):
  image_t → Backbone → Fusion → FutureState → predicted_features[t+1.6s]

Frame t+16 (actual 1.6s later, at 10Hz):
  image_{t+16} → FrozenBackbone → Fusion → target_features[t+1.6s]

feature_reconstruction_loss = MSE(predicted_features, target_features.detach())
```

### Why predict in feature space, not pixel space

Predicting future images (pixels) forces the model to reconstruct irrelevant details like exact texture, shadow positions, and compression artifacts. JEPA (Joint Embedding Predictive Architecture) [1] proposes predicting in a learned latent space instead, where only semantically meaningful changes are captured. MILE [2] demonstrated this principle specifically for autonomous driving — their latent world model outperformed pixel-prediction baselines.

### Why use a FrozenBackbone for targets

If both the predictor and the target encoder are trainable, the network can trivially minimize loss by collapsing all representations to a constant (representation collapse). Using a frozen or slowly-updating target encoder prevents this. This asymmetric design is established in BYOL [3] and JEPA [1].

Two update strategies for the frozen encoder:
1. Keep at pretrained initialization (simplest)
2. Exponential Moving Average from trainable Backbone:
   ```python
   for p_frozen, p_train in zip(frozen.parameters(), trainable.parameters()):
       p_frozen.data = τ * p_frozen.data + (1 - τ) * p_train.data  # τ ≈ 0.996
   ```

### How this helps DrivingPolicy (auxiliary task)

Training only with trajectory loss provides a weak gradient signal to the Backbone and View Fusion modules — the loss is far downstream and must propagate through many layers. Feature reconstruction loss provides a direct training signal to these early modules: "extract features that contain enough information to predict how the scene will change." This richer gradient signal improves the quality of the shared representation, which in turn benefits trajectory prediction.

UniAD [4] uses a similar multi-task design where perception auxiliary losses (detection, tracking, map segmentation) improve the shared BEV representation quality, leading to better final planning performance.

### Integration with total loss

```python
total_loss = trajectory_loss + λ * feature_reconstruction_loss
```

The weight `λ` balances the two objectives. Typical range: 0.5 – 1.0. If `λ` is too large, the model focuses on future prediction at the expense of trajectory accuracy. If too small, the auxiliary signal is negligible.

### Uncertainty estimation at inference time

After training, the feature reconstruction loss can be computed at inference time (when sequential frames are available) as a measure of model uncertainty:

```python
prediction_error = MSE(predicted_features_t_plus_1_6, actual_features_t_plus_1_6)
```

A high prediction error at a given frame indicates the model encountered a situation it did not expect — the scene evolved in a way that differs from its internal world model. This can signal:
- Unusual driving scenarios (construction zones, accidents)
- Out-of-distribution situations (weather the model has not seen)
- Sudden events (pedestrian stepping onto road)

This prediction error can be monitored as a real-time safety signal: if the model's world model is consistently wrong, the system should reduce confidence in its trajectory predictions or request human intervention.

GAIA-1 [5] demonstrates that generative world models trained on driving data develop internal representations of scene dynamics that can be used for anomaly detection in exactly this way.

## Implementation Scope

| Item | Notes |
|------|-------|
| Wire `FrozenBackbone` output as target for `FutureState` loss | Requires sequential frame pairs in DataLoader |
| Add `feature_reconstruction_loss` computation | MSE between predicted and actual features |
| Add `λ` hyperparameter | Config-level parameter |
| Add EMA update option for FrozenBackbone | Post optimizer.step() hook |
| Add prediction_error logging at inference | For introspection/monitoring |

## Out of scope (separate issues)

- Training loop infrastructure (optimizer, scheduler, DataLoader)
- Trajectory loss implementation
- Dataset selection
- BEV segmentation auxiliary loss

## References

1. LeCun, Y. "A Path Towards Autonomous Machine Intelligence." 2022. https://openreview.net/pdf?id=BZ5a1r-kVsf
2. Hu, A., et al. "Model-Based Imitation Learning for Urban Driving (MILE)." NeurIPS 2022. https://arxiv.org/abs/2210.07729
3. Grill, J.B., et al. "Bootstrap Your Own Latent (BYOL)." NeurIPS 2020. https://arxiv.org/abs/2006.07733
4. Hu, Y., et al. "Planning-oriented Autonomous Driving (UniAD)." CVPR 2023. https://arxiv.org/abs/2212.10156
5. Hu, A., et al. "GAIA-1: A Generative World Model for Autonomous Driving." 2023. https://arxiv.org/abs/2309.17080

**Comment by riita10069** [comment]
Following up on this — now that the forward pass is complete and FutureState outputs predicted future features, the missing piece to activate it is wiring up the FrozenBackbone as a target and computing the MSE between predicted and actual features.

Without this, the FutureState module runs during forward but its output is discarded — it doesn't contribute to learning, and the introspection capability (monitoring prediction error at inference time as an uncertainty signal) is also not yet functional.

I think once the trajectory loss (#15) and dataset (#16) are decided, this could be implemented relatively quickly since the FrozenBackbone already exists in the codebase. The main design questions are probably:

- Should the FrozenBackbone stay at pretrained initialization, or use EMA updates from the trainable Backbone?
- Should the loss apply to all 4 time horizons equally, or weight near-future more?

Would appreciate any thoughts from the team on priority and preferred approach. Happy to take a pass at this when the time is right.

**Comment by m-zain-khawaja** [comment]
Thanks @riita10069:
- I think we can use frozen backbone with pre-trained initialization to calculate the ground truth future state.
- For now, we can do equal weights to all for time horizons and see what the training results look like, we can then consider applying weights to specific timesteps if it helps the model converge better

**Comment by intisar1020** [comment]
I would like to  share a small incremental idea  which almost alligns with @riita10069  san proposal.

I recently came across a paper, [Temporal Difference in Vision (TDV)](https://arxiv.org/html/2606.15956v1) (just yesterday), which proposes a self-supervised learning approach that trains purely from video data using a single causal assumption -> the past causes the future. Rather than leaning on the usual inductive biases (augmentation, masking, cropping  that we see in DINO v1/v2/v3, iBOT, byol etc.), it enforces a clean additive relationship in latent space. It seems at least loosely aligned with what we're trying to do here, so I thought it might be worth a look. Following I am sharing detail.

### Core idea of TDV

The paper relies on a simple causal assumption 
 current state + encoded motion ≈ future state

Architecturally, instead of predicting absolute future features from scratch, it enforces an additive temporal relationship in latent space:

$$z_t + \Delta z_{\text{motion}} = z_{t+1}$$

where $z_t$ is the current frame representation and $\Delta z_{\text{motion}}$ is the learned temporal difference. The authors report this matches SOTA recipes on dense spatial tasks, which feels potentially relevant to our shared BEV representation goals (though I haven't verified how well the claims hold up).

### How we *might* apply this 

**1. Predicting deltas via residual learning.** Instead of having `FutureState` reconstruct absolute feature maps from scratch at each horizon, which arguably can comsume capacity re-learning static BEV geometry (roads, barriers), we could frame it as residual learning by adding the current frame's `fused_features` to the chunked `FutureState` outputs:

$$\hat{z}_{t+\Delta t} = z_t + \mathrm{FutureState}(z_t)$$

The idea is to anchor the prediction to the current spatial context, so the network mostly models what *changes* (ego-motion transitions and dynamic actors).

**2. Likely zero downstream changes.** Since the shape of $\hat{z}_{t+\Delta t}$ would stay identical to the current implementation, the loss function, training loop, and dataloader logic hopefully shouldn't need to change. We compute the standard MSE loss against the full future feature targets from the `FrozenBackbone` as @riita10069  showed earllier.

```
prediction_error = MSE(z_t + delta_motion, actual_features_t_plus_1_6)
where delta_motion = FutureState(z_t, ego_hidden)
```

**3. targets.** to compute the groudn truth latent trargets as @m-zain-khawaja  proposed we can use the frozen backbones output as it is.  The additive constraint + frozen backbone feature  could act as a good regularizer and help guard against the representation collapse usually feared in BYOL/JEPA-style setups.

### Proposed code change

The nice part is that this is a minimal, drop-in adjustment at the very end of `FutureState.forward()`, leaving the rest of the architecture untouched:

```python
def forward(self, fused_features, ego_hidden):
    # fused_features: [B, C, H, W]; ego_hidden: [B, ego_hidden_dim]

    # Inject planner intent as a per-channel bias
    bias = self.ego_proj(ego_hidden).view(-1, self.embed_dim, 1, 1)
    conditioned = fused_features + bias

    # Predict future channels
    future_features = self.predict_future_1(conditioned)
    future_features = self.activation(future_features)
    future_features = self.predict_future_2(future_features)

    # first we split into 4 future delta feature vectors where each [B, C, H, W]
    future_deltas = torch.chunk(future_features, chunks=4, dim=1)

    # Second, we apply additive residual skip connection from the current frame
    future_visual_features = tuple(fused_features + delta for delta in future_deltas)

    return future_visual_features
```


I look forward to your opinion. 


```bibtex
@misc{daithankar2026dontneedstrongassumptions,
  title         = {You Don't Need Strong Assumptions: Visual Representation Learning via Temporal Differences},
  author        = {Ninad Daithankar and Alexi Gladstone and Yann LeCun and Heng Ji},
  year          = {2026},
  eprint        = {2606.15956},
  archivePrefix = {arXiv},
  primaryClass  = {cs.CV},
  url           = {https://arxiv.org/abs/2606.15956},
}
```

**Comment by gcordova10** [comment]
Hi @m-zain-khawaja — following the priority order in #56 (riita10069 put FutureState + JEPA as #1), I've prepared the loss side of Feature B and I'd like to align with you before wiring it into training, since FutureState is your module and you have the memory-lightening action item from 17/06.

**What's ready on my side (`feat/future-state-jepa`)**
- `losses/feature_reconstruction_loss.py`: a JEPA-style feature-reconstruction loss over FutureState's predicted future feature maps, with optional temporal weighting and a zero-sum-weights guard. Additive, tested.
- Optional **action conditioning** in `future_state.py` (`action_dim=None` ⇒ byte-identical to current default).

**What I need to align with you (the "make it useful" parts riita10069 listed)**
1. **Future-frame targets** — which future features do we reconstruct (horizons +1.6/3.2/4.8/6.4 s?) and how is the target produced from the data pipeline?
2. **Target encoder** — frozen backbone vs **EMA** of the online encoder? (avoids representational collapse)
3. **Memory budget** — JEPA adds a forecast at BEV resolution (450×300). How does this fit your feature-state lightening? Options: bf16/fp16, downsample features before the forecast, fewer horizons.
4. **Loss weighting** — JEPA loss vs trajectory imitation loss (and vs any RL term).
5. **Integration point** — I assume this hooks into the training loop from #78/#67 rather than a standalone loop. Confirm?

**Proposal**: I open PR #80 with the loss + action-conditioning as the additive first step, keeping the `future_state.py` diff minimal/optional, and we iterate on (1)–(5) together. Happy to follow your lead on the memory and target-encoder design since it's your module. Thanks!

**Comment by gcordova10** [comment]
Hi @m-zain-khawaja — now that #78 and #80 are successfully merged, I’ve pushed the **target-encoder building block** (`JepaTargetEncoder`, supporting both frozen and EMA modes, stop-gradient, and `compute_jepa_loss`) to a new additive branch: `feat/t2-jepa-target-encoder`.
> It's fully tested and doesn't touch your modules. To unblock the "make JEPA useful" phase and wire this into `train_il`, we just need your lead on 6 architectural decisions:
> 1. **Target source / data:** `data_processing` currently stores `trajectory_target` but not **future frames**. To compute JEPA targets, we need future frames (or pre-extracted features) at the horizons. Should `data_processing` export these future frames, or should we re-encode them on the fly?
> 2. **Encoder Mode:** Frozen vs. EMA for the target encoder (the module supports both).
> 3. **Predictor Space:** BEV vs. feature-space (Miro mentioned "BEV/FS").
> 4. **Frequency:** Keep the JEPA branch at **1 Hz** on compressed features (leveraging your memory-lightening strategy) vs. the full 10 Hz?
> 5. **Loss Weighting:** What baseline $\lambda$ should we set for JEPA vs. the imitation loss in `train_il`?
> 6. **Memory Budget:** Does using `bf16` + 1 Hz + feature-space (avoiding the heavy BEV 450×300) align well with your 17-06 memory-lightening plan?
> 
> 
> Let me know your thoughts on these points so I can open the next PR and unblock the integration! Thanks!

**Comment by m-zain-khawaja** [comment]
Thanks @gcordova10 - regarding your specific questions:

Future-frame targets — which future features do we reconstruct (horizons +1.6/3.2/4.8/6.4 s?) and how is the target produced from the data pipeline?

- Since we have a visual history encoder currently set to 1Hz (tunable), we can have a similar 1Hz (tunable) history prediction for n frames (tunable) and the default could be 1s, 2s, 3s, 4s future horizon

Target encoder — frozen backbone vs EMA of the online encoder? (avoids representational collapse)

- Let's allow for both options

Memory budget — JEPA adds a forecast at BEV resolution (450×300). How does this fit your feature-state lightening? Options: bf16/fp16, downsample features before the forecast, fewer horizons.

- The future BEV grid is (450x300x256), predicting this over full resolution for 4 frames is very memory heavy. We can do something simple like averaging the features over the embedding dimension as per (450x300xavg(256)) -> (450x300x1) and predict over the average of the embed_dim in BEV space. However, we should allow the option to not do the averaging and predict over the full embed_dim for users who have sufficient GPU memory.

Loss weighting — JEPA loss vs trajectory imitation loss (and vs any RL term).

- for now, we can weight equally and we can see the magnitude of the losses during training and adjusting loss weighting so that each loss is approximately a similar order of magnitude

Integration point — I assume this hooks into the training loop from https://github.com/autowarefoundation/auto_e2e/pull/78/https://github.com/autowarefoundation/auto_e2e/pull/67 rather than a standalone loop

- yes, please integrate with the training loop

Feature branch

- could you please incorporate the above inputs and submit your PR to merge into the main branch

**Comment by gcordova10** [comment]
Thanks @m-zain-khawaja, this unblocks it — I'll integrate into `train_il` and open the PR to `main`. Quick design note on the memory point (3): averaging the 256-dim BEV features to a single channel hits the budget but collapses the JEPA target to an occupancy-like scalar, which loses the semantic feature-space that makes the reconstruction objective useful (I-JEPA/V-JEPA predict *abstract features* precisely to drop low-level detail — **arXiv:2301.08243**). To keep your memory budget **and** the JEPA benefit, I propose making the compression tunable on two axes instead of a fixed mean: (a) a learned channel projection `256→d_slow` (e.g. 16/32/64), and (b) optional spatial downsampling of the BEV before the predictor. A `d_slow=16` + `/4` spatial gives the same 256× memory saving as the channel-mean, but with 16 learned semantic channels instead of 1. I'd expose three modes — `occupancy` (your mean→1, cheapest baseline), `projected` (default), `full` (256, for large GPUs) — all bf16, 1 Hz, horizons 1/2/3/4 s, frozen/EMA. One open item: `data_processing` currently stores `trajectory_target` but not future frames/features — should it export the future observations at the horizons, or should `train_il` re-encode them on the fly with the (frozen/EMA) target encoder? Equal loss weights to start, with the option to switch to uncertainty weighting if one term dominates. I'll also add an optional TDV (motion-residual) predictor mode (**arXiv:2606.15956**). PR coming.

**Comment by gcordova10** [comment]
Quick status + what's left, since this issue predates the 24/06 refactor.

The objective from this proposal **is implemented — it just moved with the architecture**: the BEV `FutureState` was retired on 24/06, and the JEPA now lives in the multi-view **World Model** (#85), reusing the feature-reconstruction loss from #80 (the loss this issue asked for) and a frozen target encoder.

**Done (mapped to this proposal):**
- **Feature-reconstruction loss** → #80 (merged), in feature space, with an L1 option.
- **Frozen/EMA target encoder** + future predictor + 1 Hz history → World Model (#85).
- Your design decisions applied (@m-zain-khawaja): frozen target (EMA also supported), equal horizon weighting, 1 Hz, N=4 horizons.

**Just implemented on the training side (ready locally, tested):**
- **JEPA wired into the training loop**: `total = trajectory_loss + λ · jepa_loss`, computed via the World Model's windowed API. **λ defaults to 1.0 — equal weighting, following @m-zain-khawaja's "equal weights for now"** — and is exposed as `--jepa-weight` so it's trivial to retune once we see the two losses' magnitudes (the 0.5–1.0 range @riita10069 noted).
- **Introspection**: the per-step JEPA value is surfaced as **`prediction_error`** and logged (`world_model/prediction_error`) — the uncertainty/anomaly signal from this issue's body. (It's computed in the training loop today; the same value can be surfaced at inference for the real-time monitoring use-case.)
- Touches only `training/train.py` + tests (synthetic 1 Hz past/future windows for now), so it stays independent of @intisar1020's World Action Model work in #93.

**Still blocking the end-to-end training:**
- **Future-frame targets in the dataloader (#16)** — `data_processing` doesn't export the 1 Hz future observations yet. That's the one external dependency to train it for real; synthetic targets exercise the full two-loss step until then.

Plan: this lands as a small focused PR to `main` (`train_il` + introspection) once #85 merges. λ and the introspection logging are flags / one-liners, so they're easy to retune as training behaviour tells us more. cc @m-zain-khawaja @riita10069.

**Comment by gcordova10** [comment]
Quick status on the JEPA / World Model track (all three CI-green now that #94 is merged):

- **World Model** → **#85** (rebased on `main`, CI green): `WorldActionModel` + `AutoE2E` wiring, opt-in, with the frozen-copy target + L1 future-feature loss.
- **1 Hz dataloader** → **#95** (#16, CI green): produces the `history_frames` / `future_frames` windows the JEPA needs.
- **Training integration** (this issue): `compute_step_loss` adds `jepa_weight * jepa_loss` to the trajectory loss in `train_il` (windowed `encode_history → predict_future → jepa_loss`), with `--enable-world-model` / `--jepa-weight` flags and prediction-error introspection. It's implemented and tested on `feat/jepa-train-integration`; I'll open the PR right after #85 merges, since it builds on the `WorldActionModel` API there. Weights default to 1:1 (L1) per the 24/06 call.
